### PR TITLE
Delete `filestream` input after creating for config validation

### LINKED ISSUE #31767: `filestream` input logs an error when an existing input is reloaded with the same ID
When loading inputs from external files is enabled and there is only one `filestream` input enabled the following error is returned after the first reload:

```
filestream input with ID 'my-unique-id' already exists, this will lead to data duplication, please use a different ID
```

The problem is that the input ID stays in the bookeeper of the input manager. On reload we should remove all IDs from the manager, so the check does not interfere with previous configurations.

### Example configuration

`filebeat.yml`
```yaml
 filebeat.config.inputs:
   enabled: true
   path: inputs.d/*.yml

 output.file:
   enabled: true
```
`inputs.d/my-input.yml`
```yaml
- type: filestream
  id: my-unique-id
  paths: [/var/log/messages]
```

There is no workaround. At the moment it is not possible to configure `filestream` input from external configuration files.

Reported on discuss: https://discuss.elastic.co/t/filebeat-raise-error-filestream-input-id-already-exist

CC @belimawr

### PR DESCRIPTION
**Pull request by AndersonQ**
## What does this PR do?

Delete `filestream` input after creating for config validation

## Why is it important?

a error log (see below) is printed when the input is reloaded warning about duplicated IDs, but in reality it's the same input, therefore there is no ID duplication.

```
filestream input with ID 'my-unique-id' already exists, this will lead to data duplication, please use a different ID
```

## Checklist

- [ ] My code follows the style guidelines of this project
- ~~[ ] I have commented my code, particularly in hard-to-understand areas~~
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

follow the steps described on https://github.com/elastic/beats/issues/31767, you should not see the error log printed.

## Related issues

- Closes #31767

## Logs

TODO: add before and afeter logs

### CODE REVIEW

File: filebeat/input/filestream/internal/input-logfile/manager.go [hunk 1]
```diff
@@ -229,7 +231,19 @@ func (cim *InputManager) Create(config *conf.C) (v2.Input, error) {
 	}, nil
 }
 
-// StopInput peforms all necessary clean up when an input finishes.
+func (cim *InputManager) Delete(cfg *conf.C) error {
+	settings := struct {
+		ID string `config:"id"`
+	}{}
+	if err := cfg.Unpack(&settings); err != nil {
+		return fmt.Errorf("could not umpack conig to get inut ID: %w", err)
```
Comment (rdner) [inline filebeat/input/filestream/internal/input-logfile/manager.go:239]:
```suggestion
		return fmt.Errorf("could not unpack config to get the input ID: %w", err)
```

File: filebeat/input/filestream/internal/input-logfile/manager.go [hunk 2]
```diff
@@ -229,7 +231,19 @@ func (cim *InputManager) Create(config *conf.C) (v2.Input, error) {
...
+	}
+
+	cim.StopInput(settings.ID)
```
Comment (rdner) [inline filebeat/input/filestream/internal/input-logfile/manager.go:242]: We need to check for an empty string before this.

File: filebeat/input/filestream/internal/input-logfile/manager.go [hunk 3]
```diff
@@ -177,13 +177,15 @@ func (cim *InputManager) Create(config *conf.C) (v2.Input, error) {
...
+	cim.Logger.Infow("inputmanager ids", "ids", cim.ids)
 	cim.idsMux.Lock()
 	if _, exists := cim.ids[settings.ID]; exists {
-		cim.Logger.Errorf("filestream input with ID '%s' already exists, this "+
+		cim.Logger.Warnf("filestream input with ID '%s' already exists, this "+
 			"will lead to data duplication, please use a different ID", settings.ID)
 	}
 
 	cim.ids[settings.ID] = struct{}{}
+	cim.Logger.Infof("inputmanager added %q", settings.ID)
```
Comment (rdner) [inline filebeat/input/filestream/internal/input-logfile/manager.go:188]: we should not mentioned the input manager in INFO logs, it's internal implementation.

```suggestion
	cim.Logger.Infof("filestream with ID %q added", settings.ID)
```

File: filebeat/input/filestream/internal/input-logfile/manager.go [hunk 4]
```diff
@@ -177,13 +177,15 @@ func (cim *InputManager) Create(config *conf.C) (v2.Input, error) {
 			" duplication, please add an ID and restart Filebeat")
 	}
 
+	cim.Logger.Infow("inputmanager ids", "ids", cim.ids)
```
Comment (rdner) [inline filebeat/input/filestream/internal/input-logfile/manager.go:180]:
```suggestion
	cim.Logger.Infow("known filestream IDs", "ids", cim.ids)
```

File: filebeat/input/v2/compat/compat.go [hunk 5]
```diff
@@ -72,8 +72,15 @@ func RunnerFactory(
 func (f *factory) CheckConfig(cfg *conf.C) error {
 	_, err := f.loader.Configure(cfg)
 	if err != nil {
-		return err
+		return fmt.Errorf("runner factory could not check config: %w", err)
 	}
+
+	if err = f.loader.Delete(cfg); err != nil {
+		return fmt.Errorf(
+			"runner factory failed to delet input after checking config: %w",
```
Comment (rdner) [inline filebeat/input/v2/compat/compat.go:80]:
```suggestion
			"runner factory failed to delete an input after config check: %w",
```

File: filebeat/input/v2/input.go [hunk 6]
```diff
@@ -38,13 +38,17 @@ type InputManager interface {
 	// if inputs are only configured for testing/validation purposes.
 	Init(grp unison.Group, mode Mode) error
 
-	// Creates builds a new Input instance from the given configuation, or returns
-	// an error if the configuation is invalid.
+	// Create creates builds a new Input instance from the given configuration, or returns
```
Comment (rdner) [inline filebeat/input/v2/input.go:41]:
```suggestion
	// Create creates a new input instance from the given configuration, or returns
```

File: x-pack/filebeat/filebeat-input.yml [hunk 7]
```diff
@@ -0,0 +1,5 @@
+- type: filestream
```
Comment (rdner) [inline x-pack/filebeat/filebeat-input.yml:1]: What is this file and where is it used?

File: x-pack/filebeat/filebeat.test.yml [hunk 8]
```diff
@@ -0,0 +1,23 @@
+###################### Filebeat Configuration Example #########################
```
Comment (rdner) [inline x-pack/filebeat/filebeat.test.yml:1]: What is this file and where is it used?

**Review by rdner — CHANGES_REQUESTED** [review]
Mostly typos + I could not find tests for the new behaviour and I think it makes sense to fix the linter errors in the touched code.

File: x-pack/filebeat/filebeat-input.yml (hunk 7 above)
Comment (AndersonQ) [reply x-pack/filebeat/filebeat-input.yml:1]: sorry, dome debug stuff were left behind

File: filebeat/input/filestream/internal/input-logfile/manager.go (hunk 2 above)
Comment (rdner) [reply filebeat/input/filestream/internal/input-logfile/manager.go:240]: I don't see a point in adding the empty ID to the map.
Overriding will lead to an input running indefinitely without being closed. So, maybe we need another list to account for all the inputs then? I think this map is only valid for accounting unique IDs not the running inputs.

Comment (AndersonQ) [reply filebeat/input/filestream/internal/input-logfile/manager.go:240]: >I don't see a point in adding the empty ID to the map.

Neither do I but it's the current behaviour, so the `Delete` is just complying with it.

>I think this map is only valid for accounting unique IDs not the running inputs.

Yes, that seems to be the case.

>Overriding will lead to an input running indefinitely without being closed. So, maybe we need another list to account for all the inputs then? 

Agreed, I found it a bit odd how it works, but I did not investigate how the `ids` map is used on other parts of the code. But yes, I believe it worth investigating is there might be a bug there. However, I'd say it's for another Issue.

Comment (rdner) [reply filebeat/input/filestream/internal/input-logfile/manager.go:240]: > However, I'd say it's for another Issue.

Could you please create an issue and link it here?

**Comment by AndersonQ** [comment]
Quickly, yes, thoroughly, no.
The responsible team is working on a fix

**Comment by amitkanfer** [comment]
thanks. who's the team? can we add a team label over there?

**Comment by pierrehilbert** [comment]
According to me, the team is @elastic/obs-service-integrations

File: filebeat/input/filestream/internal/input-logfile/manager.go [hunk 9]
```diff
@@ -179,10 +179,12 @@ func (cim *InputManager) Create(config *conf.C) (v2.Input, error) {
 
 	cim.idsMux.Lock()
 	if _, exists := cim.ids[settings.ID]; exists {
-		cim.Logger.Errorf("filestream input with ID '%s' already exists, this "+
+		cim.Logger.Warnf("filestream input with ID '%s' already exists, this "+
```
Comment (belimawr) [inline filebeat/input/filestream/internal/input-logfile/manager.go:182]: This is still an error. It is a common user error when writing a standalone Filebeat configuration and it does lead to data duplication.

Unfortunately we cannot make an unique ID required as it is a breaking change, so the log error is the best way to tell our users their configuration is going to cause trouble.

File: filebeat/input/v2/compat/compat.go [hunk 10]
```diff
@@ -146,7 +153,11 @@ func configID(config *conf.C) (string, error) {
 	}
 
 	var h map[string]interface{}
-	config.Unpack(&h)
+	err := config.Unpack(&h)
+	if err != nil {
+		return "", fmt.Errorf("can not compute id from configuration: unpack failed: %w", err)
```
Comment (belimawr) [inline filebeat/input/v2/compat/compat.go:158]: There are other reasons `config.Unpack` could fail here:
```suggestion
		return "", fmt.Errorf("cannot parse configuration: unpack failed: %w", err)
```

**Review by belimawr — CHANGES_REQUESTED** [review]
It would be nice to have an integration test that tests the case described on https://github.com/elastic/beats/issues/31767. We could also go a step further and test we can reload inputs with the same ID (not only start them like on https://github.com/elastic/beats/issues/31767's example).

That can easily be done with a system tests and the input reload feature enabled.

**Comment by AndersonQ** [comment]
> It would be nice to have an integration test that tests the case described on #31767. We could also go a step further and test we can reload inputs with the same ID (not only start them like on #31767 example).
> 
> That can easily be done with a system tests and the input reload feature enabled.

@belimawr, there is already a unit test https://github.com/elastic/beats/pull/35134/files#diff-76df5c54e03464b2c7ca44fb58060212afe6bf0c88bc66c45eb9bbda453bc596R166. 

Why do you think an integration test is better, and at what level, which components would be in the integration test? Or are you referring to an E2E test?

File: filebeat/input/v2/compat/compat.go (hunk 10 above)
Comment (AndersonQ) [reply filebeat/input/v2/compat/compat.go:158]: It does not change the fact that the ID could not be extracted from the configuration. That way whoever reads the error know what was tried and the root cause, thus having more information to understand what is happening.

File: filebeat/input/filestream/internal/input-logfile/manager.go (hunk 2 above)
Comment (AndersonQ) [reply filebeat/input/filestream/internal/input-logfile/manager.go:242]: Thus an empty ID id as valid as any other and suffer from the same uniqueness problem.
@rdner do you agree?

Comment (rdner) [reply filebeat/input/filestream/internal/input-logfile/manager.go:242]: @AndersonQ I don't really understand your comment but I trust @belimawr on this one, he dealt with that in the past.

**Review by rdner — APPROVED** [review]

**Comment by belimawr** [comment]
> Why do you think an integration test is better, and at what level, which components would be in the integration test? Or are you referring to an E2E test?

The whole Filebeat, I meant the Python Integration Tests that you can run with `mage pythonIntegTest` and live in `https://github.com/elastic/beats/tree/main/filebeat/tests/system`.

Because it will test not only the `InputManager` but whomever is creating and discarding inputs to validate configuration like the `compat.go`: https://github.com/elastic/beats/blob/d4a196c044452db962928a718c40d9ab256a04dc/filebeat/input/v2/compat/compat.go#L72-L85

The unit test you pointed ensures the `InputManager` works, however it does not ensure the users of `InputManager` actually call the new `Delete` method.

Now that I'm thinking more about it, I believe we should have either `InputManage.Delete` or `InputManager.StopInput`. `InputManage.Delete` seems to be a sugar syntax to get the ID then call `InputManager.StopInput`

**Comment by cmacknz** [comment]
+1 to Tiago's suggestion, let's make sure we an integration test that covers this change before merge.

**Review by belimawr — APPROVED** [review]

**Comment by belimawr** [comment]
@AndersonQ it seems you missed the backport to `v8.8.0`, I added the tag, could you merge the backport PR once CI is green?

@fixed77 thanks for spotting that! The fix should be released in `v8.8.1`.

**Comment by AndersonQ** [comment]
thanks

**Pull request merged**